bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/i18n/ThingTypeI18nLocalizationService.java

```diff
@@ -65,29 +72,67 @@ protected void unsetChannelTypeRegistry(ChannelTypeRegistry channelTypeRegistry)
-                thingType.getDescription(), locale);
+    @Nullable
+    private StateDescription createLocalizedStateDescription(Bundle bundle, ChannelType channelType,
+            ChannelTypeUID channelTypeUID, Locale locale) {
+        StateDescription state = channelType.getState();
+
+        if (state != null) {
+            String pattern = thingTypeI18nUtil.getChannelStatePattern(bundle, channelTypeUID, state.getPattern(),
+                    locale);
+
+            List<StateOption> localizedOptions = new ArrayList<>();
+            List<StateOption> options = state.getOptions();
+            for (StateOption stateOption : options) {
+                String optionLabel = thingTypeI18nUtil.getChannelStateOption(bundle, channelTypeUID,
+                        stateOption.getValue(), stateOption.getLabel(), locale);
+                localizedOptions.add(new StateOption(stateOption.getValue(), optionLabel));
+            }
+
+            return new StateDescription(state.getMinimum(), state.getMaximum(), state.getStep(), pattern,
+                    state.isReadOnly(), localizedOptions);
+        }
+        return null;
+    }
+
+    public ChannelType createLocalizedChannelType(Bundle bundle, ChannelType channelType, Locale locale) {
+        ChannelTypeUID channelTypeUID = channelType.getUID();
+        String label = thingTypeI18nUtil.getChannelLabel(bundle, channelTypeUID, channelType.getLabel(), locale);
+        String description = thingTypeI18nUtil.getChannelDescription(bundle, channelTypeUID,
+                channelType.getDescription(), locale);
 
-        final List<ChannelDefinition> localizedChannelDefinitions = new ArrayList<>(
-                thingType.getChannelDefinitions().size());
+        StateDescription state = createLocalizedStateDescription(bundle, channelType, channelTypeUID, locale);
+
+        return new ChannelType(channelTypeUID, channelType.isAdvanced(), channelType.getItemType(),
+                channelType.getKind(), label, description, channelType.getCategory(), channelType.getTags(), state,
+                channelType.getEvent(), channelType.getConfigDescriptionURI());
+    }
 
-        for (final ChannelDefinition channelDefinition : thingType.getChannelDefinitions()) {
-            String channelLabel = this.thingTypeI18nUtil.getChannelLabel(bundle, thingType.getUID(), channelDefinition,
+    public ChannelGroupType createLocalizedChannelGroupType(Bundle bundle, ChannelGroupType channelGroupType,
+            Locale locale) {
+        ChannelGroupTypeUID channelGroupTypeUID = channelGroupType.getUID();
+        String label = thingTypeI18nUtil.getChannelGroupLabel(bundle, channelGroupTypeUID, channelGroupType.getLabel(),
+                locale);
+        String description = thingTypeI18nUtil.getChannelGroupDescription(bundle, channelGroupTypeUID,
+                channelGroupType.getDescription(), locale);
+
+        List<ChannelDefinition> localizedChannelDefinitions = new ArrayList<>(
```

Comment:
The lines 119-142 are almost identical to 154-176 but I could not find a way to extract a method for them. Any thoughts?

Reply:
Difficult to identify the problem in the Github diff view only. What's the problem?

Reply:
@maggu2810 Thanks for your review.

This is the duplicate code which I want to extract into a new method. The problem is that I want to determine the `channelLabel` and `channelDescription` inside the foreach loop derived from either the `thingTypeUID` or the `channelGroupTypeUID` - I marked the code with **.

```java
    private List<ChannelDefinition> createLocalizedChannelDefinitions(Bundle bundle,
            List<ChannelDefinition> channelDefinitions, Locale locale) {
        // start
        List<ChannelDefinition> localizedChannelDefinitions = new ArrayList<>(channelDefinitions.size());
        for (ChannelDefinition channelDefinition : channelDefinitions) {
            String channelLabel = thingTypeI18nUtil.getChannelLabel(bundle, **<thingTypeUID|channelGroupTypeUID>**, channelDefinition,
                    channelDefinition.getLabel(), locale);
            String channelDescription = thingTypeI18nUtil.getChannelDescription(bundle, **<thingTypeUID|channelGroupTypeUID>**, channelDefinition,
                    channelDefinition.getDescription(), locale);
            if (channelLabel == null || channelDescription == null) {
                ChannelTypeUID channelTypeUID = channelDefinition.getChannelTypeUID();
                ChannelType channelType = channelTypeRegistry.getChannelType(channelTypeUID, locale);
                if (channelType != null) {
                    if (channelLabel == null) {
                        channelLabel = thingTypeI18nUtil.getChannelLabel(bundle, channelTypeUID, channelType.getLabel(),
                                locale);
                    }
                    if (channelDescription == null) {
                        channelDescription = thingTypeI18nUtil.getChannelDescription(bundle, channelTypeUID,
                                channelType.getDescription(), locale);
                    }
                }
            }
            localizedChannelDefinitions
                    .add(new ChannelDefinition(channelDefinition.getId(), channelDefinition.getChannelTypeUID(),
                            channelDefinition.getProperties(), channelLabel, channelDescription));
        }
        // end
        return localizedChannelDefinitions;
    }
```

Reply:
```diff
diff --git a/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/i18n/ThingTypeI18nLocalizationService.java b/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/i18n/ThingTypeI18nLocalizationService.java
index df5c1c14fb..2a2b50c766 100644
--- a/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/i18n/ThingTypeI18nLocalizationService.java
+++ b/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/i18n/ThingTypeI18nLocalizationService.java
@@ -15,6 +15,7 @@ package org.eclipse.smarthome.core.thing.i18n;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Locale;
+import java.util.function.Function;
 
 import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.eclipse.jdt.annotation.Nullable;
@@ -116,31 +117,12 @@ public class ThingTypeI18nLocalizationService {
         String description = thingTypeI18nUtil.getChannelGroupDescription(bundle, channelGroupTypeUID,
                 channelGroupType.getDescription(), locale);
 
-        List<ChannelDefinition> localizedChannelDefinitions = new ArrayList<>(
-                channelGroupType.getChannelDefinitions().size());
-        for (ChannelDefinition channelDefinition : channelGroupType.getChannelDefinitions()) {
-            String channelLabel = thingTypeI18nUtil.getChannelLabel(bundle, channelGroupTypeUID, channelDefinition,
-                    channelDefinition.getLabel(), locale);
-            String channelDescription = thingTypeI18nUtil.getChannelDescription(bundle, channelGroupTypeUID,
-                    channelDefinition, channelDefinition.getDescription(), locale);
-            if (channelLabel == null || channelDescription == null) {
-                ChannelTypeUID channelTypeUID = channelDefinition.getChannelTypeUID();
-                ChannelType channelType = channelTypeRegistry.getChannelType(channelTypeUID, locale);
-                if (channelType != null) {
-                    if (channelLabel == null) {
-                        channelLabel = thingTypeI18nUtil.getChannelLabel(bundle, channelTypeUID, channelType.getLabel(),
-                                locale);
-                    }
-                    if (channelDescription == null) {
-                        channelDescription = thingTypeI18nUtil.getChannelDescription(bundle, channelTypeUID,
-                                channelType.getDescription(), locale);
-                    }
-                }
-            }
-            localizedChannelDefinitions
-                    .add(new ChannelDefinition(channelDefinition.getId(), channelDefinition.getChannelTypeUID(),
-                            channelDefinition.getProperties(), channelLabel, channelDescription));
-        }
+        List<ChannelDefinition> localizedChannelDefinitions = localizeChannelDefinitions(bundle, channelGroupType.getChannelDefinitions(),
+                channelDefinition -> thingTypeI18nUtil.getChannelLabel(bundle, channelGroupTypeUID, channelDefinition,
+                        channelDefinition.getLabel(), locale),
+                channelDefinition -> thingTypeI18nUtil.getChannelDescription(bundle, channelGroupTypeUID,
+                        channelDefinition, channelDefinition.getDescription(), locale),
+                locale);
 
         return new ChannelGroupType(channelGroupTypeUID, channelGroupType.isAdvanced(), label, description,
                 channelGroupType.getCategory(), localizedChannelDefinitions);
@@ -151,30 +133,12 @@ public class ThingTypeI18nLocalizationService {
         String label = thingTypeI18nUtil.getLabel(bundle, thingTypeUID, thingType.getLabel(), locale);
         String description = thingTypeI18nUtil.getDescription(bundle, thingTypeUID, thingType.getDescription(), locale);
 
-        List<ChannelDefinition> localizedChannelDefinitions = new ArrayList<>(thingType.getChannelDefinitions().size());
-        for (ChannelDefinition channelDefinition : thingType.getChannelDefinitions()) {
-            String channelLabel = thingTypeI18nUtil.getChannelLabel(bundle, thingTypeUID, channelDefinition,
-                    channelDefinition.getLabel(), locale);
-            String channelDescription = thingTypeI18nUtil.getChannelDescription(bundle, thingTypeUID, channelDefinition,
-                    channelDefinition.getDescription(), locale);
-            if (channelLabel == null || channelDescription == null) {
-                ChannelTypeUID channelTypeUID = channelDefinition.getChannelTypeUID();
-                ChannelType channelType = channelTypeRegistry.getChannelType(channelTypeUID, locale);
-                if (channelType != null) {
-                    if (channelLabel == null) {
-                        channelLabel = thingTypeI18nUtil.getChannelLabel(bundle, channelTypeUID, channelType.getLabel(),
-                                locale);
-                    }
-                    if (channelDescription == null) {
-                        channelDescription = thingTypeI18nUtil.getChannelDescription(bundle, channelTypeUID,
-                                channelType.getDescription(), locale);
-                    }
-                }
-            }
-            localizedChannelDefinitions
-                    .add(new ChannelDefinition(channelDefinition.getId(), channelDefinition.getChannelTypeUID(),
-                            channelDefinition.getProperties(), channelLabel, channelDescription));
-        }
+        List<ChannelDefinition> localizedChannelDefinitions = localizeChannelDefinitions(bundle, thingType.getChannelDefinitions(),
+                channelDefinition -> thingTypeI18nUtil.getChannelLabel(bundle, thingTypeUID, channelDefinition,
+                        channelDefinition.getLabel(), locale),
+                channelDefinition -> thingTypeI18nUtil.getChannelDescription(bundle, thingTypeUID, channelDefinition,
+                        channelDefinition.getDescription(), locale),
+                locale);
 
         List<ChannelGroupDefinition> localizedChannelGroupDefinitions = new ArrayList<>(
                 thingType.getChannelGroupDefinitions().size());
@@ -213,4 +177,32 @@ public class ThingTypeI18nLocalizationService {
         return builder.build();
     }
 
+    public <T> List<ChannelDefinition> localizeChannelDefinitions(final Bundle bundle, final List<ChannelDefinition> channelDefinitions,
+            final Function<ChannelDefinition, String> channelLabelResolver,
+            final Function<ChannelDefinition, String> channelDescriptionResolver, final Locale locale) {
+        List<ChannelDefinition> localizedChannelDefinitions = new ArrayList<>(channelDefinitions.size());
+        for (ChannelDefinition channelDefinition : channelDefinitions) {
+            String channelLabel = channelLabelResolver.apply(channelDefinition);
+            String channelDescription = channelDescriptionResolver.apply(channelDefinition);
+            if (channelLabel == null || channelDescription == null) {
+                ChannelTypeUID channelTypeUID = channelDefinition.getChannelTypeUID();
+                ChannelType channelType = channelTypeRegistry.getChannelType(channelTypeUID, locale);
+                if (channelType != null) {
+                    if (channelLabel == null) {
+                        channelLabel = thingTypeI18nUtil.getChannelLabel(bundle, channelTypeUID, channelType.getLabel(),
+                                locale);
+                    }
+                    if (channelDescription == null) {
+                        channelDescription = thingTypeI18nUtil.getChannelDescription(bundle, channelTypeUID,
+                                channelType.getDescription(), locale);
+                    }
+                }
+            }
+            localizedChannelDefinitions
+                    .add(new ChannelDefinition(channelDefinition.getId(), channelDefinition.getChannelTypeUID(),
+                            channelDefinition.getProperties(), channelLabel, channelDescription));
+        }
+        return localizedChannelDefinitions;
+    }
+
 }
```

Reply:
Very nice. Works like a charm. Thank you very much.

Reply:
Now I will leave it to another committer to press the merge button :wink: